### PR TITLE
Support properties system table in Hive

### PR DESCRIPTION
**Pull request by qqibrow**
Support Query syntax `SELECT * FROM "table_name$properties" ` to return properties defined in  hive metastore

Ref https://github.com/prestodb/presto/pull/10793
Fixes https://github.com/prestodb/presto/issues/10611

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/security/SystemTableAwareAccessControl.java [hunk 1]
```diff
@@ -40,6 +40,90 @@ public SystemTableAwareAccessControl(ConnectorAccessControl delegate)
...
+    {
+        delegate.checkCanDropColumn(transactionHandle, identity, tableName);
+    }
+
+    @Override
+    public void checkCanRenameColumn(ConnectorTransactionHandle transactionHandle, ConnectorIdentity identity, SchemaTableName tableName)
+    {
+        delegate.checkCanRenameColumn(transactionHandle, identity, tableName);
+    }
+
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/security/SystemTableAwareAccessControl.java]: these above methods should be already defined in `ForwardingConnectorAccessControl`, no need to have them here

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 2]
```diff
@@ -332,15 +334,51 @@ public ConnectorTableHandle getTableHandleForStatisticsCollection(ConnectorSessi
...
 
-    private Optional<SystemTable> getPartitionsSystemTable(ConnectorSession session, SchemaTableName tableName)
+    private Optional<SystemTable> getPropertiesSystemTable(ConnectorSession connectorSession, SchemaTableName tableName, SchemaTableName sourceTableName)
+    {
+        Optional<Table> table = metastore.getTable(sourceTableName.getSchemaName(), sourceTableName.getTableName());
+        if (!table.isPresent() || table.get().getTableType().equals(TableType.VIRTUAL_VIEW.name())) {
+            throw new TableNotFoundException(tableName);
+        }
+        List<Map.Entry<String, String>> sortedEntries = table.get().getParameters().entrySet().stream().sorted(Comparator.comparing(Map.Entry::getKey)).collect(toImmutableList());
+        List<ColumnMetadata> columns = sortedEntries.stream().map(Map.Entry::getKey).map(key -> new ColumnMetadata(key, VarcharType.VARCHAR)).collect(toImmutableList());
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: See my comment on https://github.com/prestodb/presto/issues/10611

Using separate columns per parameter is the reason we have a separate system table per Hive table. It makes it easy to use interactively:
```sql
SELECT location FROM "logs$properties"
```
vs
```sql
SELECT value FROM "logs$properties" WHERE name = 'location'
```

We can also implement a global table of key/value pairs.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 3]
```diff
@@ -2194,16 +2232,32 @@ public void commit()
...
     {
-        return tableName.getTableName().endsWith(PARTITIONS_TABLE_SUFFIX) && tableName.getTableName().length() > PARTITIONS_TABLE_SUFFIX.length();
+        return Stream.of(PARTITIONS_TABLE_HANDLER, PROPERTIES_TABLE_HANDLER)
+                .filter(handler -> handler.matches(tableName))
+                .map(handler -> handler.getSourceTableName(tableName))
+                .findAny();
     }
 
-    public static SchemaTableName getSourceTableNameForPartitionsTable(SchemaTableName tableName)
+    private interface SystemTableHandler
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: That's a good idea. It could look like
```java
private enum SystemTableHandler
{
    PARTITIONS, PROPERTIES;

    private final String suffix;

    SystemTableHandler()
    {
        this.suffix = "$" + name().toLowerCase(ENGLISH);
    }

    public boolean matches(...)

    SchemaTableName getSourceTableName(...)
}

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 4]
```diff
@@ -332,15 +334,51 @@ public ConnectorTableHandle getTableHandleForStatisticsCollection(ConnectorSessi
...
+    private Optional<SystemTable> getPropertiesSystemTable(ConnectorSession connectorSession, SchemaTableName tableName, SchemaTableName sourceTableName)
+    {
+        Optional<Table> table = metastore.getTable(sourceTableName.getSchemaName(), sourceTableName.getTableName());
+        if (!table.isPresent() || table.get().getTableType().equals(TableType.VIRTUAL_VIEW.name())) {
+            throw new TableNotFoundException(tableName);
+        }
+        List<Map.Entry<String, String>> sortedEntries = table.get().getParameters().entrySet().stream().sorted(Comparator.comparing(Map.Entry::getKey)).collect(toImmutableList());
+        List<ColumnMetadata> columns = sortedEntries.stream().map(Map.Entry::getKey).map(key -> new ColumnMetadata(key, VarcharType.VARCHAR)).collect(toImmutableList());
+        List<Type> types = columns.stream().map(ColumnMetadata::getType).collect(toImmutableList());
+        Iterable<List<Object>> propertyValues = ImmutableList.of(ImmutableList.copyOf(sortedEntries.stream().map(Map.Entry::getValue).collect(toImmutableList())));
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: This could be simplified to
```java
Map<String, String> sortedTableParameters = ImmutableSortedMap.copyOf(table.get().getParameters());
List<ColumnMetadata> columns = sortedTableParameters.entrySet().stream()
```
Use a sorted map rather than `List<Entry>`

**Comment by dain** [comment]
@cla-bot check

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveConnectorFactory.java [hunk 5]
```diff
@@ -27,7 +27,7 @@
 import io.prestosql.plugin.hive.metastore.HiveMetastoreModule;
 import io.prestosql.plugin.hive.s3.HiveS3Module;
 import io.prestosql.plugin.hive.security.HiveSecurityModule;
-import io.prestosql.plugin.hive.security.PartitionsAwareAccessControl;
+import io.prestosql.plugin.hive.security.SystemTableAwareAccessControl;
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveConnectorFactory.java]: We should actually keep this rename in the same commit, since the behavior is changing along with the name. Doing it before or after would be misleading, as the behavior would not match the name.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 6]
```diff
@@ -332,15 +334,51 @@ public ConnectorTableHandle getTableHandleForStatisticsCollection(ConnectorSessi
...
+
+        SystemTable propertiesSystemTable = new SystemTable()
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java]: Let's change this a bit, then we can use it for both system tables:
```java
private static SystemTable createSystemTable(ConnectorTableMetadata metadata, Function<TupleDomain<Integer>, RecordCursor> cursor)
{
    return new SystemTable()
    {
        @Override
        public Distribution getDistribution()
        {
            return Distribution.SINGLE_COORDINATOR;
        }

        @Override
        public ConnectorTableMetadata getTableMetadata()
        {
            return metadata;
        }

        @Override
        public RecordCursor cursor(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
        {
            return cursor.apply(constraint);
        }
    }
}
```
The partitions usage would look something like
```java
return Optional.of(createSystemTable(
        new ConnectorTableMetadata(tableName, partitionSystemTableColumns),
        constraint -> {
            TupleDomain<ColumnHandle> targetTupleDomain = constraint.transform(fieldIdToColumnHandle::get);
            Predicate<Map<ColumnHandle, NullableValue>> targetPredicate = convertToPredicate(targetTupleDomain);
            Constraint<ColumnHandle> targetConstraint = new Constraint<>(targetTupleDomain, targetPredicate);
            Iterable<List<Object>> records = () ->
                    stream(partitionManager.getPartitions(metastore, sourceTableHandle, targetConstraint).getPartitions())
                            .map(hivePartition ->
                                    IntStream.range(0, partitionColumns.size())
                                            .mapToObj(fieldIdToColumnHandle::get)
                                            .map(columnHandle -> hivePartition.getKeys().get(columnHandle).getValue())
                                            .collect(toList()))
                            .iterator();

            return new InMemoryRecordSet(partitionColumnTypes, records).cursor();
        }));
```

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 7]
```diff
@@ -590,10 +592,32 @@ private void testCreateTableAs(Session session, HiveStorageFormat storageFormat)
     @Test
     public void testCreatePartitionedTableAs()
     {
-        testWithAllStorageFormats(this::testCreatePartitionedTableAs);
+        testWithAllStorageFormats(this::createPartitionedTableAs);
```
Comment (electrum) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: This rename seems unrelated. Please revert it.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 8]
```diff
@@ -590,10 +592,32 @@ private void testCreateTableAs(Session session, HiveStorageFormat storageFormat)
...
 
-    private void testCreatePartitionedTableAs(Session session, HiveStorageFormat storageFormat)
+    @Test
+    public void testPropertiesTable()
+    {
+        @Language("SQL") String createTable = "" +
+                "CREATE TABLE test_show_properties" +
+                " WITH (" +
+                "format = 'orc', " +
+                "partitioned_by = ARRAY[ 'SHIP_PRIORITY', 'ORDER_STATUS' ]," +
```
Comment (electrum) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: Nit: don't put spaces around array brackets. Also, use lowercase for the column names.
```sql
partitioned_by = ARRAY['ship_priority', 'order_status'],
orc_bloom_filter_columns = ARRAY['ship_priority', 'order_status'],

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 9]
```diff
@@ -590,10 +592,32 @@ private void testCreateTableAs(Session session, HiveStorageFormat storageFormat)
...
+                "orc_bloom_filter_columns = ARRAY[ 'SHIP_PRIORITY', 'ORDER_STATUS' ]," +
+                "orc_bloom_filter_fpp = 0.5" +
+                ") " +
+                "AS " +
+                "SELECT orderkey AS order_key, shippriority AS ship_priority, orderstatus AS order_status " +
+                "FROM tpch.tiny.orders";
+
+        assertUpdate(createTable, "SELECT count(*) FROM orders");
+        BasicQueryInfo createTableQuery = ((DistributedQueryRunner) getQueryRunner()).getCoordinator().getQueryManager().getQueries().stream().filter(basicQueryInfo -> basicQueryInfo.getQuery().equals(createTable)).findFirst().get();
```
Comment (electrum) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: This is a clever way to get the query ID and node version! However, there is a simpler way:
```java
String queryId = (String) computeScalar("SELECT query_id FROM system.runtime.queries WHERE query LIKE 'CREATE TABLE test_show_properties%'");
String nodeVersion = (String) computeScalar("SELECT node_version FROM system.runtime.nodes WHERE coordinator");

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 10]
```diff
@@ -590,10 +592,32 @@ private void testCreateTableAs(Session session, HiveStorageFormat storageFormat)
...
+        String nodeVersion = ((DistributedQueryRunner) getQueryRunner()).getCoordinator().refreshNodes().getActiveNodes().stream().map(Node::getVersion).findAny().get();
+        assertQuery("SELECT * FROM \"test_show_properties$properties\"", "SELECT '" + "ship_priority,order_status" + "','" + "0.5" + "','" + createTableQuery.getQueryId().getId() + "','" + nodeVersion + "'");
```
Comment (electrum) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: This is hard to read. Please format like
```java
assertQuery(
        "SELECT * FROM \"test_show_properties$properties\"",
        format("SELECT 'ship_priority,order_status', 0.5, '%s', '%s', queryId, nodeVersion));

**Comment by electrum** [comment]
Merged, thanks!
